File: HW2.py
'''
HW2
python code for homework2 in computer1 class in Yonsei University
Using numerical integration to solve Kepler problem
'''
from typing import Tuple
import numpy as np
import matplotlib.pyplot as plt
from HW1 import HW1


def HW2(zeta_min: float, zeta_max: float,
        n: int) -> Tuple[np.ndarray, np.ndarray]:

    '''
    Integrate :math:`\\zeta\'(0.987654(\\zeta\'-0.9)(9/8-\\zeta\'))^{-1/2}`
    from 0.9 to :math:`\\zeta`
    To remove singularity in the integrand, I separates the integrand
    and changes
    :math:`\\zeta\'` to :math:`x=(\\zeta\'-0.9)^{1/2}` for the first part and
    to :math:`x=(9/8-\\zeta\')^{1/2}` for the second part.

    Args:
     zeta_min: minimum value of zeta
     zeta_max: maximum value of zeta
     n: number of points between zeta_min and zeta_max

    Returns:
     Integration evaluated in n points between zeta_min and zeta_max
    '''

    zeta = np.linspace(zeta_min, zeta_max, n+1)
    t = np.zeros(n+1)
    x1 = np.sqrt(zeta-zeta_min)
    x2 = np.sqrt(zeta_max-zeta)
    t[0] = 0
    for i in range(n):
        t[i+1] = t[i] + 8/np.sqrt(0.987654)*(x1[i+1]-x1[i])*x2[i] - \
            10/np.sqrt(0.987654)*(x2[i+1]-x2[i])*x1[i]

    return t, zeta


if __name__ == "__main__":

    t_ref, zeta_ref = HW1(0, 3.2, 1000, 0.9, 0)
    t, zeta = HW2(0.9, 9/8, 10)
    plt.plot(t_ref, zeta_ref, label='HW1')
    plt.plot(t, zeta, marker='o', markersize=3,
             linestyle='none', label='HW2')
    plt.legend()
    plt.show()

File: HW2.cpp
/*
HW2
C++ code for homework2 in computer1 class in Yonsei University
Using numerical integration to solve Kepler problem
*/

#include <cmath>
#include <tuple>
#include <vector>
#include <iostream>

using namespace std;

tuple<vector<double>, vector<double>> HW2(double a,
					  double t0,
					  double zeta_min, double zeta_max,
					  int n)
{

  //initialize variable
  double grid_space;
  vector<double> zeta(n+1, 0);
  vector<double> t(n+1, 0);
  vector<double> x1(n+1, 0);
  vector<double> x2(n+1, 0);

  // use uniform n pts between zeta_min and zeta_max
  // additional one point needed for end point.

  grid_space = (zeta_max-zeta_min)/n;
  
  for(int i=0; i<n+1; i++)
    {
      zeta[i] = i*grid_space + zeta_min;
    }

  // change of variable
  // and seperate integral
  for(int i=0; i<n+1; i++)
    {
      x1[i] = sqrt(zeta[i]-zeta_min);
      x2[i] = sqrt(zeta_max-zeta[i]);
    }

  // initial condtion zeta(t0) = zeta_min
  t[0] = t0;

  // numerical integration using Riemann-Stieltiges integral
  // Quite fun, integrand_1(x1) = x2 and
  // integrand_2(x2) = x1
  for(int i=1; i<n+1; i++)
    {
      t[i] = t[i-1] + 8/sqrt(a)*(x1[i]-x1[i-1])*x2[i-1] - \
	10/sqrt(a)*(x2[i]-x2[i-1])*x1[i-1];
    }

  return make_tuple(t, zeta);
}

int main(void) {
  double zeta_min = 0.9;
  double zeta_max = 9./8.;
  int n = 20;
  double t0 = 0;
  double a = 0.987654;
  vector<double> t, y;
  
  tie(t, y) = HW2(a, t0, zeta_min, zeta_max, n);

    for (int i = 0; i < n+1; i++)
      {
        cout << t[i] << '\t' << y[i] << endl;
      }

    return 0;
}

  
     
			    
			  
    
